api/src/operations/exec/index.ts

```diff
@@ -0,0 +1,44 @@
+import { defineOperationApi, toArray } from '@directus/shared/utils';
+import { NodeVM, NodeVMOptions, VMScript } from 'vm2';
+
+type Options = {
+	code: string;
+};
+
+export default defineOperationApi<Options>({
+	id: 'exec',
+	handler: async ({ code }, { data, env }) => {
+		const allowedModules = env.FLOWS_EXEC_ALLOWED_MODULES ? toArray(env.FLOWS_EXEC_ALLOWED_MODULES) : [];
+
+		const opts: NodeVMOptions = {
+			eval: false,
+			wasm: false,
+		};
+
+		if (allowedModules.length > 0) {
+			opts.require = {
+				external: {
+					modules: allowedModules,
+					transitive: false,
+				},
+			};
+		}
+
+		const vm = new NodeVM(opts);
+
+		let script;
+
+		try {
+			script = new VMScript(code).compile();
+		} catch (err: any) {
+			throw new Error(`Couldn't compile code: ${err?.message}`);
+		}
+
+		try {
+			const fn = await vm.run(script);
+			return await fn(data);
+		} catch (err: any) {
+			throw new Error(`Couldn't run code: ${err?.message}`);
+		}
```

Comment:
@rijkvanzanten The `payload` for a reject flow is simply `{}` when an `Error` type object is propagated.

```suggestion
		const fn = await vm.run(script);
		return await fn(data);
```

With the above change, different kinds of errors will be propagated upstream.
Note: The unit tests will have to be updated accordingly.

We will also have to update the following: 

https://github.com/directus/directus/blob/f5778047c1825d995b54e81f2c2a2f43f6ad4e48/api/src/flows.ts#L397

to: 
```ts
return {
	successor: operation.reject,
	status: 'reject',
	data: (error ? (error.message ? error.message : error) : error) ?? null,
	options,
};
```

Reply:
I was partially wondering what was the cases where there's no `.message`, as it's `err?.message`. But yeah in the context of this operation in particular, I reckon the error should be propagated so that the reject operations can access the error thrown and proceed with different logics accordingly. 👍

Reply:
Yeah I totally forgot that the whole exec is ran inside a try/catch _anyways_, so it's a little pointless to re-throw errors here 🙂